File: src/main/java/com/ptaf/ai/audit/GenerationAuditSupport.java
package com.ptaf.ai.audit;

import com.ptaf.ai.config.AiAssistantProperties;
import com.ptaf.ai.model.AiGenerationMode;
import com.ptaf.ai.model.GenerationResult;

import java.nio.file.Path;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

/** Builds and appends generation audit records for CLI and HTTP entry points. */
public final class GenerationAuditSupport {

    private GenerationAuditSupport() {
    }

    public static List<String> collectWarnings(GenerationResult result) {
        if (result == null) {
            return List.of();
        }
        List<String> warnings = new ArrayList<>();
        warnings.addAll(result.structuredResponse().warnings());
        if (result.stepReuseValidationResult() != null) {
            warnings.addAll(result.stepReuseValidationResult().warnings());
        }
        if (result.yamlKeyValidationResult() != null) {
            warnings.addAll(result.yamlKeyValidationResult().warnings());
        }
        if (result.allowedYamlGuardResult() != null) {
            warnings.addAll(result.allowedYamlGuardResult().warnings());
        }
        if (result.pageFrameContextGuardResult() != null) {
            warnings.addAll(result.pageFrameContextGuardResult().warnings());
        }
        if (result.runnableFeatureResult() != null) {
            warnings.addAll(result.runnableFeatureResult().warnings());
        }
        if (result.missingYamlPatchSuggestions() != null) {
            for (var suggestion : result.missingYamlPatchSuggestions()) {
                warnings.addAll(suggestion.warnings());
            }
        }
        return warnings;
    }

    public static AiGenerationAuditRecord buildRecord(
            AiGenerationMode mode,
            AiAssistantProperties props,
            String requirement,
            Path requestedOutput,
            Path writtenOutput,
            GenerationResult result,
            List<String> blockingErrors,
            List<String> warnings
    ) {
        boolean parseOk = result != null && result.structuredResponse() != null && result.structuredResponse().parseSuccessful();
        boolean stepOk = result != null
                && result.stepReuseValidationResult() != null
                && result.stepReuseValidationResult().passed();
        boolean yamlOk = result != null
                && result.yamlKeyValidationResult() != null
                && result.yamlKeyValidationResult().passed();
        int reused = result != null && result.structuredResponse() != null ? result.structuredResponse().reusedSteps().size() : 0;
        int newSteps = result != null && result.structuredResponse() != null ? result.structuredResponse().newStepsNeeded().size() : 0;
        int missingYaml = result != null && result.yamlKeyValidationResult() != null ? result.yamlKeyValidationResult().missingCount() : 0;
        String outputPath = writtenOutput != null ? writtenOutput.toString() : requestedOutput.toString();

        return new AiGenerationAuditRecord(
                UUID.randomUUID().toString(),
                Instant.now().toString(),
                "generate",
                mode.name(),
                props.model(),
                props.promptVersion(),
                AiGenerationAuditLogger.sha256(requirement),
                outputPath,
                parseOk,
                stepOk,
                yamlOk,
                writtenOutput != null,
                reused,
                newSteps,
                missingYaml,
                warnings,
                blockingErrors
        );
    }

    public static AiGenerationAuditLogger.WriteResult append(
            Path projectRoot,
            AiAssistantProperties props,
            AiGenerationMode mode,
            String requirement,
            Path requestedOutput,
            Path writtenOutput,
            GenerationResult result,
            List<String> blockingErrors
    ) {
        List<String> warnings = collectWarnings(result);
        AiGenerationAuditRecord record = buildRecord(
                mode,
                props,
                requirement,
                requestedOutput,
                writtenOutput,
                result,
                blockingErrors,
                warnings
        );
        return new AiGenerationAuditLogger().append(
                projectRoot,
                props.auditEnabled(),
                props.auditOutputPath(),
                record
        );
    }
}

File: python/ptaf/ai/audit/generation_audit_support.py
from __future__ import annotations

from datetime import datetime, timezone
from pathlib import Path
from uuid import uuid4

from ptaf.ai.audit.ai_generation_audit_logger import AiGenerationAuditLogger
from ptaf.ai.audit.ai_generation_audit_record import AiGenerationAuditRecord
from ptaf.ai.config.ai_assistant_properties import AiAssistantProperties
from ptaf.ai.model.ai_generation_mode import AiGenerationMode
from ptaf.ai.model.generation_result import GenerationResult


def collect_warnings(result: GenerationResult | None) -> list[str]:
    if result is None:
        return []
    warnings: list[str] = []
    warnings.extend(result.structured_response.warnings)
    if result.step_reuse_validation_result:
        warnings.extend(result.step_reuse_validation_result.warnings)
    if result.yaml_key_validation_result:
        warnings.extend(result.yaml_key_validation_result.warnings)
    if result.allowed_yaml_guard_result:
        warnings.extend(result.allowed_yaml_guard_result.warnings)
    if result.page_frame_context_guard_result:
        warnings.extend(result.page_frame_context_guard_result.warnings)
    if result.runnable_feature_result:
        warnings.extend(result.runnable_feature_result.warnings)
    if result.missing_yaml_patch_suggestions:
        for suggestion in result.missing_yaml_patch_suggestions:
            warnings.extend(suggestion.warnings)
    return warnings


def build_record(
    mode: AiGenerationMode,
    props: AiAssistantProperties,
    requirement: str,
    requested_output: Path,
    written_output: Path | None,
    result: GenerationResult | None,
    blocking_errors: list[str],
    warnings: list[str],
) -> AiGenerationAuditRecord:
    parse_ok = result is not None and result.structured_response.parse_successful
    step_ok = (
        result is not None
        and result.step_reuse_validation_result is not None
        and result.step_reuse_validation_result.passed
    )
    yaml_ok = (
        result is not None
        and result.yaml_key_validation_result is not None
        and result.yaml_key_validation_result.passed
    )
    reused = len(result.structured_response.reused_steps) if result else 0
    new_steps = len(result.structured_response.new_steps_needed) if result else 0
    missing_yaml = result.yaml_key_validation_result.missing_count if result and result.yaml_key_validation_result else 0
    output_path = str(written_output) if written_output else str(requested_output)

    return AiGenerationAuditRecord(
        request_id=str(uuid4()),
        timestamp=datetime.now(timezone.utc).isoformat(),
        command="generate",
        mode=mode.name,
        model=props.model(),
        prompt_version=props.prompt_version(),
        requirement_hash=AiGenerationAuditLogger.sha256(requirement),
        output_path=output_path,
        parse_successful=parse_ok,
        step_validation_passed=step_ok,
        yaml_validation_passed=yaml_ok,
        file_written=written_output is not None,
        reused_steps_count=reused,
        new_steps_count=new_steps,
        missing_yaml_keys_count=missing_yaml,
        warnings=warnings,
        blocking_errors=blocking_errors,
    )


def append(
    project_root: Path,
    props: AiAssistantProperties,
    mode: AiGenerationMode,
    requirement: str,
    requested_output: Path,
    written_output: Path | None,
    result: GenerationResult | None,
    blocking_errors: list[str],
) -> AiGenerationAuditLogger.WriteResult:
    warnings = collect_warnings(result)
    record = build_record(
        mode,
        props,
        requirement,
        requested_output,
        written_output,
        result,
        blocking_errors,
        warnings,
    )
    return AiGenerationAuditLogger().append(
        project_root,
        props.audit_enabled(),
        props.audit_output_path(),
        record,
    )
